From b0880dc00682a231ae2b96450bcfe0f1846fed9d Mon Sep 17 00:00:00 2001
From: Graham Whaley <graham.whaley@intel.com>
Date: Thu, 8 Dec 2016 16:17:53 +0000
Subject: [PATCH] 9p: fix create->unlink->fstat bug

Patch originally by Eric Van Hensbergen, at:
https://github.com/ericvh/linux-kvm/commit/2fa2f7e728ac08a7d9006516870db1a986aa6acc

Rebased to lkvm repo (non-full-Kernel) as used by rkt/stage1/kvm.

Original commit message:
this is the kvmtool side of the bug fix.  Essentially if we have
a fid that is already open then use fstat instead of lstat to avoid
the aforementioned problem.

Signed-off-by: Graham Whaley <graham.whaley@intel.com>
---
 virtio/9p.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/virtio/9p.c b/virtio/9p.c
index 6acbfdd..ce368c7 100644
--- a/virtio/9p.c
+++ b/virtio/9p.c
@@ -677,8 +677,14 @@ static void virtio_p9_getattr(struct p9_dev *p9dev,
 
 	virtio_p9_pdu_readf(pdu, "dq", &fid_val, &request_mask);
 	fid = get_fid(p9dev, fid_val);
-	if (lstat(fid->abs_path, &st) < 0)
-		goto err_out;
+
+	if(fid->fd > 0) {
+		if(fstat(fid->fd, &st) < 0)
+			goto err_out;
+	} else {
+		if (lstat(fid->abs_path, &st) < 0)
+			goto err_out;
+	}
 
 	virtio_p9_fill_stat(p9dev, &st, &statl);
 	virtio_p9_pdu_writef(pdu, "A", &statl);
-- 
2.7.4

